### ISSUE TITLE
same-origin request definition around A->B->A redirects

### ISSUE THREAD

**Issue by hiroshige-g** [w3c/webappsec-referrer-policy#128]
(please correct my understanding if I missed something)

https://w3c.github.io/webappsec-referrer-policy/#same-origin-request

> A Request request is a same-origin request if request’s origin and the origin of request’s current url are the same.

When a page on `https://a.example.com` origin sends a request to `https://b.example.com/first` which is then redirected to `https://a.example.com/final`, the final request is considered as a "same-origin request" according to the definition above, because `https://a.example.com/final` is same-origin as the page.
Therefore, if referrer policy `"origin-when-cross-origin"` is set, the full URL (rather than the origin) is sent as the `Referer` header of the request to `https://a.example.com/final`, according to Step 7 of
https://fetch.spec.whatwg.org/#concept-main-fetch

However, I feel this kind of A->B->A redirecting requests should be considered cross-origin-ish

- To align with CORS etc. (which rejects A->B->A requests in `same-origin` mode)
- As all the browsers also use the origin as `Referer` in the case above, as tested in:
https://wpt.fyi/results/referrer-policy/gen/top.meta/origin-when-cross-origin/fetch/cross-http.swap-origin.http.html?label=experimental&label=master&aligned

and the same-origin definition should be modified, while I'm not so sure how (there can be multiple, slightly different ways to fix this case), and I feel this is a Pri-3 editorial issue.
WDYT?

**Comment by domfarolino** [comment]
> Therefore, if referrer policy "origin-when-cross-origin" is set, the full URL (rather than the origin) is sent as the Referer header of the request to https://a.example.com/final, according to Step 7 of
https://fetch.spec.whatwg.org/#concept-main-fetch

OK I could also be missing something here, but when the request is first sent from A -> B, the request's referrer string would be stripped to the origin given the referrer policy. Then when went redirected from B -> A, this is considered same-origin by the spec, but even so, the full referrer string could be at most the already-stripped origin, right? Is this undesirable? Would it be beneficial to take into consideration a request's [tainted origin flag](https://fetch.spec.whatwg.org/#concept-request-tainted-origin)?

**Comment by hiroshige-g** [comment]
> the full referrer string could be at most the already-stripped origin, right?

Oh, I see, I thought as if the referrer is recalculated from the scratch for each redirection, but in fact we first set the original `request's referrer` when the initial fetch is started (usually with `"client"` case of https://w3c.github.io/webappsec-referrer-policy/#determine-requests-referrer, which is never used at redirects), then strip `request's referrer` down iteratively if needed.

So now I think the spec's results match with my expectation.
I don't have further motivations to refactor the mechanism e.g. based on tainted origin flag.

Thanks for clarification! Closing.
